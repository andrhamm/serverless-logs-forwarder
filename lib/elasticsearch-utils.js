import elasticsearch from 'elasticsearch';

export const client = opts => new elasticsearch.Client({
  // log: 'trace', // uncomment for debug logging
  apiVersion: '5.6',
  protocol: 'http',
  requestTimeout: 6000,
  ...opts,
});
