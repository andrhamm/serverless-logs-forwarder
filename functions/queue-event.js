import { sns } from '../lib/aws-clients';

const { SNS_TOPIC_ARN_LOGGROUP_EVENTS } = process.env;

export const handler = async (event) => {
  await sns.publish({
    Message: JSON.stringify(event),
    TopicArn: SNS_TOPIC_ARN_LOGGROUP_EVENTS,
  }).promise();
};
