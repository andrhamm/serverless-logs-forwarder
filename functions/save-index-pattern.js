/* eslint-disable no-console, import/no-extraneous-dependencies */
import { getEnv } from '../lib/common';
import { getStackNameForLogGroup } from '../lib/cloudformation-utils';
import { saveIndexPattern } from '../lib/kibana-utils';

let env;
export const handler = async (sqsEvent) => {
  if (!env) {
    env = await getEnv([
      'KIBANA_API_BASE',
      'LOG_GROUP_PREFIX',
      'STAGE',
    ]);
  }

  let event;
  if (sqsEvent.Records) {
    event = JSON.parse(sqsEvent.Records[0].body);
  } else {
    event = sqsEvent;
  }

  // eg. /aws/lambda/logging-demo-dev-api
  let logGroupName = event.logGroupName;
  if (!logGroupName) {
    logGroupName = event.detail.requestParameters.logGroupName;

    if (env.LOG_GROUP_PREFIX && !logGroupName.startsWith(env.LOG_GROUP_PREFIX)) {
      console.log(`ignoring the log group [${logGroupName}] because it doesn't match the prefix [${env.LOG_GROUP_PREFIX}]`);
      return;
    }
  }

  const stackName = await getStackNameForLogGroup(logGroupName);

  if (stackName) {
    await saveIndexPattern({
      stackName,
      stage: env.STAGE,
      kibanaApiBase: env.KIBANA_API_BASE,
    });

    console.log(`updated [${logGroupName}]'s kibana index pattern`);
  } else {
    console.log(`no stack for [${logGroupName}], skipping kibana index pattern`);
  }
};
