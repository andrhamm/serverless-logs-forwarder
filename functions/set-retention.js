/* eslint-disable no-console, import/no-extraneous-dependencies */
import { getEnv } from '../lib/common';
import { getExpiry, setExpiry } from '../lib/cloudwatch-utils';

let env;
export const handler = async (sqsEvent) => {
  if (!env) {
    env = await getEnv([
      'KINESIS_DESTINATION_ARN',
      'LOG_GROUP_PREFIX',
      'RETENTION_DAYS',
      'ROLE_ARN',
    ]);
  }

  let event;
  if (sqsEvent.Records) {
    event = JSON.parse(sqsEvent.Records[0].body);
  } else {
    event = sqsEvent;
  }

  // eg. /aws/lambda/logging-demo-dev-api
  let logGroupName = event.logGroupName;
  if (!logGroupName) {
    logGroupName = event.detail.requestParameters.logGroupName;

    if (env.LOG_GROUP_PREFIX && !logGroupName.startsWith(env.LOG_GROUP_PREFIX)) {
      console.log(`ignoring the log group [${logGroupName}] because it doesn't match the prefix [${env.LOG_GROUP_PREFIX}]`);
      return;
    }
  }

  if (!(await getExpiry(logGroupName))) {
    await setExpiry({ logGroupName, retentionInDays: env.RETENTION_DAYS });
    console.log(`updated [${logGroupName}]'s retention policy to ${env.RETENTION_DAYS} days`);
  }
};
