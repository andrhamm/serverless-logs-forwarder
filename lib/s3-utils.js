import { tryParseJson } from './common';

import { s3 } from '../lib/aws-clients';

export const tryGetObjectJSON = async (params) => {
  let mappingObject;
  try {
    mappingObject = await s3.getObject(params).promise();
  } catch (e) {
    if (e.code !== 'NoSuchKey') {
      throw e;
    }
  }

  let mapping = {};
  if (mappingObject && mappingObject.Body) {
    const decoded = tryParseJson(mappingObject.Body);

    if (decoded) mapping = decoded;
  }

  return mapping;
};

export const putObjectJSON = (params, mapping) => s3.putObject({
  Body: JSON.stringify(mapping),
  ContentType: 'application/json',
  ...params,
}).promise();
