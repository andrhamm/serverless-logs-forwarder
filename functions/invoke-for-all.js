/* eslint-disable no-console, import/no-extraneous-dependencies */
import AWS from 'aws-sdk'; // this dep is present in Lambda environment automatically
import pMap from 'p-map';
import { getEnv } from '../lib/common';
import { tryGetObjectJSON } from '../lib/s3-utils';

const lambda = new AWS.Lambda({ apiVersion: '2015-03-31' });

let env;

export const handler = async (event, context, callback) => {
  if (!env) {
    env = await getEnv([
      'S3_BUCKET_NAME',
      'LAMBDA_FUNCTION_NAME_PREFIX',
    ]);
  }

  const mappingParams = {
    Bucket: env.S3_BUCKET_NAME,
    Key: 'cloudwatch-cloudformation-mapping.json',
  };

  const mapping = await tryGetObjectJSON(mappingParams);
  const logGroups = Object.keys(mapping);

  const { lambdaFunctionName } = event;

  if (!lambdaFunctionName) {
    callback('Event missing \'lambdaFunctionName\' property');
    return;
  }

  const fullFunctionName = lambdaFunctionName.startsWith(env.LAMBDA_FUNCTION_NAME_PREFIX) ?
    lambdaFunctionName : `${env.LAMBDA_FUNCTION_NAME_PREFIX}${lambdaFunctionName}`;

  console.log(`Async invoking '${fullFunctionName}' function for ${logGroups.length} log groups`);

  await pMap(logGroups, logGroupName => lambda.invoke({
    FunctionName: fullFunctionName,
    InvocationType: 'Event',
    Payload: JSON.stringify({ logGroupName }),
  }).promise(), { concurrency: 4 });
};
