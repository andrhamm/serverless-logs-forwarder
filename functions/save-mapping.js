/* eslint-disable no-console */
import { getEnv } from '../lib/common';
import { tryGetObjectJSON, putObjectJSON } from '../lib/s3-utils';
import { getLogGroupStackMapping } from '../lib/cloudformation-utils';

let env;

export const handler = async () => {
  if (!env) {
    env = await getEnv([
      'LOG_GROUP_PREFIX',
      'S3_BUCKET_NAME',
    ]);
  }

  const mappingParams = {
    Bucket: env.S3_BUCKET_NAME,
    Key: 'cloudwatch-cloudformation-mapping.json',
  };

  const [mapping, prevMapping] = await Promise.all([
    getLogGroupStackMapping({ logGroupNamePrefix: env.LOG_GROUP_PREFIX }),
    tryGetObjectJSON(mappingParams),
  ]);

  const newMapping = {
    ...prevMapping,
    ...mapping,
  };

  console.log(`Saving mapping for ${Object.keys(newMapping).length} log groups`);

  await putObjectJSON(mappingParams, newMapping);
};
