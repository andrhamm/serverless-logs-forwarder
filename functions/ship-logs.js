import { getEnv } from '../lib/common';
import { extractLogs } from '../lib/parse';
import * as elasticsearch from '../lib/elasticsearch-utils';
import { tryGetObjectJSON } from '../lib/s3-utils';

let mapping;
let env;
let es;

export const handler = async (event) => {
  if (!env) {
    env = await getEnv([
      'ELASTICSEARCH_HOST',
      'ELASTICSEARCH_HTTPAUTH_ENCRYPTED',
      'S3_BUCKET_NAME',
    ]);
  }

  const mappingParams = {
    Bucket: env.S3_BUCKET_NAME,
    Key: 'cloudwatch-cloudformation-mapping.json',
  };

  let mappingPromise;
  if (!mapping) {
    mappingPromise = tryGetObjectJSON(mappingParams);
  }

  const logs = extractLogs(event.Records);
  delete event.Records;

  const dt = new Date();
  const indexDate = [
    dt.getFullYear(),
    (`0${dt.getMonth() + 1}`).slice(-2),
    (`0${dt.getDate()}`).slice(-2),
  ].join('.');

  if (mappingPromise) mapping = await mappingPromise;

  const body = logs.reduce((acc, log) => {
    const stackOrServiceName = mapping[log.logGroup] || 'lambda';

    acc.push({
      index: {
        _index: `cloudwatchlogs-${stackOrServiceName}-${indexDate}`,
        _type: 'cloudwatchlog',
        _id: log.id,
      },
    });

    delete log.id;

    acc.push(log);

    return acc;
  }, []);

  if (body.length > 0) {
    if (!es) {
      es = elasticsearch.client({
        host: env.ELASTICSEARCH_HOST,
        httpAuth: env.ELASTICSEARCH_HTTPAUTH,
      });
    }

    await es.bulk({ body });
  }
};
