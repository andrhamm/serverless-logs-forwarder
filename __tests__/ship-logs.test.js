import * as elasticsearch from '../lib/elasticsearch-utils';
import * as s3Utils from '../lib/s3-utils';
import { kinesisifyRecord, mockKinesisCloudWatchLogsEventRecords } from './__helpers__/kinesis';
import { handler } from '../functions/ship-logs';

jest.mock('../lib/elasticsearch-utils');
jest.mock('../lib/s3-utils');

afterEach(() => { jest.resetAllMocks(); });

describe('ship-logs', () => {
  describe('handler', () => {
    it('calls elasticsearch bulk', async () => {
      expect.assertions(1);

      const mockMapping = mockKinesisCloudWatchLogsEventRecords.reduce((acc, r, i) => {
        if (!acc[r.logGroup]) {
          acc[r.logGroup] = `stack-${i}`;
        }

        return acc;
      }, {});

      const tryGetObjectJSON = jest.spyOn(s3Utils, 'tryGetObjectJSON');
      tryGetObjectJSON.mockResolvedValue(mockMapping);

      const client = jest.spyOn(elasticsearch, 'client');
      const esBulkFn = jest.fn();

      client.mockReturnValue({
        bulk: esBulkFn,
      });

      await handler({
        Records: mockKinesisCloudWatchLogsEventRecords.map(kinesisifyRecord),
      });

      expect(esBulkFn.mock.calls[0]).toMatchSnapshot();
    });
  });
});
