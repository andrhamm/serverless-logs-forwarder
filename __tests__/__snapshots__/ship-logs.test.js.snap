// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ship-logs handler calls elasticsearch bulk 1`] = `
Array [
  Object {
    "body": Array [
      Object {
        "index": Object {
          "_id": "1",
          "_index": "cloudwatchlogs-stack-0-2018.11.06",
          "_type": "cloudwatchlog",
        },
      },
      Object {
        "@timestamp": 2015-05-28T15:27:35.000Z,
        "functionName": "my-stack-my_function",
        "lambdaVersion": "$LATEST",
        "level": "info",
        "logGroup": "/aws/lambda/my-stack-my_function",
        "logStream": "2018/11/05/[$LATEST]a3a206b7b35f414f9f567242385ede8a",
        "message": "message 1",
        "request_id": 1,
        "type": "cloudwatch",
      },
      Object {
        "index": Object {
          "_id": "2",
          "_index": "cloudwatchlogs-stack-0-2018.11.06",
          "_type": "cloudwatchlog",
        },
      },
      Object {
        "@timestamp": 2015-05-28T15:27:35.000Z,
        "functionName": "my-stack-my_function",
        "lambdaVersion": "$LATEST",
        "level": "info",
        "logGroup": "/aws/lambda/my-stack-my_function",
        "logStream": "2018/11/05/[$LATEST]a3a206b7b35f414f9f567242385ede8a",
        "message": "message 2",
        "request_id": 2,
        "type": "cloudwatch",
      },
      Object {
        "index": Object {
          "_id": "3",
          "_index": "cloudwatchlogs-stack-0-2018.11.06",
          "_type": "cloudwatchlog",
        },
      },
      Object {
        "@timestamp": 2015-05-28T15:27:35.000Z,
        "fields": Object {
          "customField": "custom field value",
        },
        "functionName": "my-stack-my_function",
        "lambdaVersion": "$LATEST",
        "level": "info",
        "logGroup": "/aws/lambda/my-stack-my_function",
        "logStream": "2018/11/05/[$LATEST]a3a206b7b35f414f9f567242385ede8b",
        "message": null,
        "request_id": 3,
        "type": "cloudwatch",
      },
      Object {
        "index": Object {
          "_id": "4",
          "_index": "cloudwatchlogs-stack-0-2018.11.06",
          "_type": "cloudwatchlog",
        },
      },
      Object {
        "@timestamp": 2015-05-28T15:27:35.000Z,
        "functionName": "my-stack-my_function",
        "lambdaVersion": "$LATEST",
        "level": "info",
        "logGroup": "/aws/lambda/my-stack-my_function",
        "logStream": "2018/11/05/[$LATEST]a3a206b7b35f414f9f567242385ede8b",
        "message": "message 4",
        "request_id": 4,
        "type": "cloudwatch",
      },
    ],
  },
]
`;
